src/main/java/ch/njol/skript/expressions/ExprItemsIn.java

```diff
@@ -45,104 +47,126 @@
  */
 @Name("Items In")
 @Description({"All items in an inventory. Useful for looping or storing in a list variable.",
-		"Please note that the positions of the items in the inventory are not saved, only their order is preserved."})
+	"Please note that the positions of the items in the inventory are not saved, only their order is preserved."})
 @Examples({"loop all items in the player's inventory:",
-		"	loop-item is enchanted",
-		"	remove loop-item from the player",
-		"set {inventory::%uuid of player%::*} to items in the player's inventory"})
+	"	loop-item is enchanted",
+	"	remove loop-item from the player",
+	"set {inventory::%uuid of player%::*} to items in the player's inventory"})
 @Since("2.0")
 public class ExprItemsIn extends SimpleExpression<Slot> {
 	static {
-		Skript.registerExpression(ExprItemsIn.class, Slot.class, ExpressionType.PROPERTY, "[(all [[of] the]|the)] items ([with]in|of|contained in|out of) (|1¦inventor(y|ies)) %inventories%");
+		Skript.registerExpression(ExprItemsIn.class, Slot.class, ExpressionType.PROPERTY, "[(all [[of] the]|the)] (items|%-itemtypes%) ([with]in|of|contained in|out of) (|1¦inventor(y|ies)) %inventories%");
 	}
-	
+
 	@SuppressWarnings("null")
 	private Expression<Inventory> invis;
-	
+
+	@Nullable
+	private Expression<ItemType> types;
+
 	@SuppressWarnings({"unchecked", "null"})
 	@Override
 	/*
 	 * the parse result will be null if it is used via the ExprInventory expression, however the expression will never
 	 * be a variable when used with that expression (it is always a anonymous SimpleExpression)
 	 */
 	public boolean init(final Expression<?>[] exprs, final int matchedPattern, final Kleenean isDelayed, @Nullable final ParseResult parseResult) {
-		invis = (Expression<Inventory>) exprs[0];
+		types = (Expression<ItemType>) exprs[0];
+		invis = (Expression<Inventory>) exprs[1];
 		if (invis instanceof Variable && !invis.isSingle() && parseResult.mark != 1)
 			Skript.warning("'items in {variable::*}' does not actually represent the items stored in the variable. Use either '{variable::*}' (e.g. 'loop {variable::*}') if the variable contains items, or 'items in inventories {variable::*}' if the variable contains inventories.");
 		return true;
 	}
-	
+
+	private boolean isAllowedItem(@Nullable ItemType[] types, @Nullable ItemStack item) {
+		if (types == null)
+			return item != null;
+		else if (item == null)
+			return false;
+
+		ItemType potentiallyAllowedItem = new ItemType(item);
+		for (ItemType type : types) {
+			if (potentiallyAllowedItem.isSimilar(type)) {
+				return true;
+			}
+		}
+
+		return false;
+	}
+
 	@SuppressWarnings("null")
 	@Override
 	protected Slot[] get(final Event e) {
 		final ArrayList<Slot> r = new ArrayList<>();
+		final ItemType[] types = this.types == null ? null : this.types.getArray(e);
```

Comment:
What should the behaviour for `all (stone or dirt) in {_inv}` be? Because currently, it'll return either only the stone or only the dirt, chosen randomly.

Reply:
I think that the current behavior you described is probably OK